src/Service/ProductPositionsUpdater.php

```diff
@@ -0,0 +1,117 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use ThreeBRS\SortingPlugin\Service\Exception\ProductPositionsUpdaterException;
+
+class ProductPositionsUpdater
+{
+    private ?FlashBagInterface $sessionBag = null;
+
+    /**
+     * @var array<int|string, int> Key: PK, Value: index
+     */
+    private array $identifiersPositionsMap = [];
+
+    public function __construct(
+        private EntityManagerInterface $entityManager,
+        private RouterInterface $router,
+        private EventDispatcherInterface $eventDispatcher,
+        private TranslatorInterface $translator,
+    )
+    {
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    public function process(Request $request): string
+    {
+        $this->initialize($request);
+
+        if ($this->identifiersPositionsMap === []) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.chooseTaxon');
+        }
+
+        $productTaxonCollection = $this->entityManager
+            ->getRepository(ProductTaxonInterface::class)
+            ->findBy(['id' => array_keys($this->identifiersPositionsMap)]);
+
+        if (count($productTaxonCollection) !== count($this->identifiersPositionsMap)) {
+            throw new ProductPositionsUpdaterException('Required identifiers set is invalid.');
+        }
+
+        $taxon = null;
+
+        foreach ($productTaxonCollection as $productTaxon) {
+            assert($productTaxon instanceof ProductTaxonInterface);
+            $taxon ??= $productTaxon->getTaxon();
+            $productTaxon->setPosition($this->identifiersPositionsMap[$productTaxon->getId()] ?? 0);
+        }
+
+        if ($taxon === null) {
+            $this->setSessionError('threebrs.ui.sortingPlugin.noProductMessage');
+
+            return $this->router->generate('threebrs_admin_sorting_index');
+        }
+
+        $this->entityManager->flush();
+        $this->setSessionSuccess('threebrs.ui.sortingPlugin.successMessage');
+        $this->eventDispatcher->dispatch(new GenericEvent($taxon), 'threebrs-sorting-products-after-persist');
+
+        return $this->router->generate('threebrs_admin_sorting_products', ['taxonId' => $taxon->getId()]);
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function prepareSessionMessage(string $key): string
+    {
+        if ($this->sessionBag === null) {
+            throw new ProductPositionsUpdaterException('Session\'s state is corrupted.');
+        }
+
+        return $this->translator->trans($key);
+    }
+
+    /**
+     * @throws ProductPositionsUpdaterException
+     */
+    private function setSessionError(string $messageKey): void
+    {
+        $this->sessionBag->add(
+            'error',
+            $this->prepareSessionMessage($messageKey)
```

Comment:
překlad bych vyřešil zde `$this->translator->trans($key)`

Reply:
translator nevyhazuje chybu pokud nenajde překlad, jen vypíše klíč. takže voláš metodu která volá metodu, je to zbytečně navíc

Reply:
Nepoužívám github, takže přiznávám že nevím co mám kde potvrdit.

Našel jsem pouze _Re-request_ review, což jsem před chvílí použil.

Reply:
Radši se zeptám ještě jednou - opravdu tam nechcete try catch? Protože Translator může vyhodit `InvalidArgumentException` v `\Symfony\Component\Translation\Translator::assertValidLocale`, `RuntimeException` v `\Symfony\Component\Translation\Translator::doLoadCatalogue` a jediná situace, kterou si to řeší samo, je `\Symfony\Component\Translation\Exception\NotFoundResourceException` což je to co píšeš.

Takže když někdo pošle locale s překlepem, nebo selže načítání překladů ze zdroje, tak padne celá response.

Nechám na tobě, je nezvyklá situace, co se tu snažím řešit, to uznávám. Ale taky vím, že nezvyklé situace, které nemá framework pořešené, mě stojí nejvíc času vyřešit.